### PR TITLE
Load history data asynchronously, as separate bundles per package card

### PR DESCRIPTION
**Pull request by NickGerleman**
This moves from synchronously loading a bundle with all download history, to separating the download history for each package into a separate bundle, and async loading it. This allows us to start showing UI when ready (using multiple parallel downloads), instead of waiting on a single very large bundle to be downloaded and parsed.

I don't force sequential fade-in of cards, but it doesn't seem too jarring. Will possibly followup with a change to force sequential order.

https://user-images.githubusercontent.com/835219/147048840-b0a9a91c-1635-4813-93d6-db267d472272.mp4

.

### CODE REVIEW

File: src/HistoryReader.ts [hunk 1]
```diff
@@ -25,21 +22,49 @@ export default class HistoryReader {
...
+    packageIdentifier: PackageIdentifier
+  ): Promise<HistoryReader> {
     if (!HistoryReader.instances[packageIdentifier]) {
-      HistoryReader.instances[packageIdentifier] = new HistoryReader(
+      const historyFile = await HistoryReader.loadHistoryFile(
         packageIdentifier
       );
+
+      //throw new Error(Array.isArray(historyFile).toString());
+
```
Comment (NickGerleman) [inline src/HistoryReader.ts:42]:
```suggestion
```

File: scripts/generateAssets.ts [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+    const historyAssetPath = path.join(
+      __dirname,
+      "..",
+      "src/generated_assets",
+      `${packageName.replace(/\//g, "_")}.json`
+    );
+
+    const historyFile: AssetHistoryFile = { points: sortedPoints };
+    await fs.writeFile(historyAssetPath, JSON.stringify(historyFile));
```
Comment (tido64) [inline scripts/generateAssets.ts:64]: Do we really need to await here?

**Review by tido64 — APPROVED** [review]
Doesn't React Suspense support this scenario? I thought it was explicitly designed for it, but now it looks like we're manually animating in the sections now.

File: scripts/generateAssets.ts (hunk 2 above)
Comment (NickGerleman) [reply scripts/generateAssets.ts:64]: We import the promise based version of `fs` above:

```ts
import { promises as fs } from "fs";
```

We could write concurrently to give up sequentiality, but I'd kind of rather this be sequential, to be easier to reason about.

**Comment by NickGerleman** [comment]
> Doesn't React Suspense support this scenario? I thought it was explicitly designed for it, but now it looks like we're manually animating in the sections now.

I was using suspense before. I don't think it mapped well to the scenario of loading separate data bundles, to be used with the same component. I think React.Lazy is a bit better for loading a single component, and its dependencies.

**Comment by NickGerleman** [comment]
Total downloaded is 3.4MB

**Comment by NickGerleman** [comment]
We see benefits with concurrent downloads when we have slow connections relative to latency, E.g. the artificially throttled version below.
image

File: scripts/generateAssets.ts (hunk 2 above)
Comment (tido64) [reply scripts/generateAssets.ts:64]: Sorry, what I meant is that I think you can fire-and-forget in this case. It doesn't look like the rest of the logic needs to wait till the write is done.

**Comment by tido64** [comment]
> > Doesn't React Suspense support this scenario? I thought it was explicitly designed for it, but now it looks like we're manually animating in the sections now.
> 
> I was using suspense before. I don't think it mapped well to the scenario of loading separate data bundles, to be used with the same component. I think React.Lazy is a bit better for loading a single component, and its dependencies.

It should support this scenario. I've seen React demos where separate sections of a page are fetching data from different sources, e.g. avatars or loading data based on priority. Maybe this is something that comes in React 18?

**Comment by NickGerleman** [comment]
> > > Doesn't React Suspense support this scenario? I thought it was explicitly designed for it, but now it looks like we're manually animating in the sections now.
> > 
> > 
> > I was using suspense before. I don't think it mapped well to the scenario of loading separate data bundles, to be used with the same component. I think React.Lazy is a bit better for loading a single component, and its dependencies.
> 
> It should support this scenario. I've seen React demos where separate sections of a page are fetching data from different sources, e.g. avatars or loading data based on priority. Maybe this is something that comes in React 18?

React as of 17 (what this is targeting atm) includes `React.lazy`, documented to currently support lazy async importing a bundle for a specific component, and showing a Suspense component in the meantime.

I must admit I haven't looked at what is new with 18 and concurrent mode.

**Comment by NickGerleman** [comment]
This is using the same dynamic import as `React.lazy` would though, where we allow webpack to split a bunde based on boundaries we create in code. We split data in this case though, instead of the bundle from the component.

**Comment by tido64** [comment]
> I must admit I haven't looked at what is new with 18 and concurrent mode.

Yeah, I was thinking of [SuspenseList](https://reactjs.org/docs/concurrent-mode-patterns.html#suspenselist) (the name had completely slipped my mind). It gives you loading state and controlled reveal, which is what we want here. But it's not available till 18.

**Pull request merged**